.github/workflows/cd.yaml

```diff
@@ -4,9 +4,28 @@ on:
   push:
     branches:
       - master
+      - eks-test-2
 
 jobs:
+  pre-clean:
+    # Delete existing S3 folder and K8s namespace
+    # if they already exist from a previous run.
+    runs-on: ubuntu-latest
+    steps:
+      - uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: us-east-2
+      - run: |
+          set -x
+          ZQD_DATA_URI=s3://zq-ci-test/$(date +%y-%m-%d).${{ github.run_number }}
+          aws eks --region us-east-2 update-kubeconfig --name zq-test
+          aws s3 rm --recursive $ZQD_DATA_URI || :
+          kubectl delete ns ci-${{ github.run_number }} || :
   ecr-image-push:
+    # Build and push a Docker image for zqd.
+    needs: pre-clean
```

Comment:
This file would be shorter (and clearer) if you used a single job rather than a sequence of jobs linked with `needs`. For instance, you'd only need one `uses: aws-actions/configure-aws-credentials@v1` step instead of four.

Reply:
It is important that we can clearly distinguish between the jobs that fail and succeed. Combining the jobs would make the "happy path" easier to read, but would give us information about critical failures more slowly, and would require us to dig into the output to understand what failed.

Each job has a name that corresponds to its independent goal: clean, image build, deploy, test, master deploy. This code is optimized for operational clarity, as it should be. Making the code shorter is less important that the clarity in the operation of each job.

Reply:
> Combining the jobs would make the "happy path" easier to read, but would give us information about critical failures more slowly, and would require us to dig into the output to understand what failed.

I don't think that's true, but the argument's tough to make in words, so let's merge this as is and explore later.

Reply:
> Combining the jobs would make the "happy path" easier to read, but would give us information about critical failures more slowly, and would require us to dig into the output to understand what failed.

By this reasoning, you could say that we should break up every step in the "test" job (ie make fmt, make tidy, make test-unit, make test-system etc) into its own job would make it easier to identify where the failure occurred. But if we did this we would have so many jobs the list would be huge.  Fwiw I haven't had an issue with finding the reason for a failure in "test" job.

If this was called "eks-smoke test" and it failed while everything else passed, I would immediately have in my mind the parts of my pr that could have caused the failure. Having a small number of logically grouped together test jobs makes it easier to have a clear picture of what your ci test infrastructure looks like. If you a bunch of narrowly defined tests, you place a higher burden on institutional knowledge.

I don't want to hold up this pr, but I would prefer this was structured as a single job and moreover to keep an eye on the number of jobs we are creating.

Reply:
The reasoning is different from what you describe. This is not a single test. It is a collection of the most important external integration points with the deployment technologies we are using. If the Docker build fails, that is a much more significant failure than a unit test: it means our code is not be being built on our selected deployment platform. The deploy step failing means that our AWS infrastructure cannot accept what we built, and EKS may be misconfigured, or our K8s yaml is broken. This is a very different type of failure than a functional test. The test step in this will contain an extensive set of distributed tests, reflecting a subset of our local ZTests, and need not be broken into multiple jobs.